keanu-project/src/main/java/io/improbable/keanu/plating/Plate.java

```diff
@@ -1,23 +1,58 @@
 package io.improbable.keanu.plating;
 
-import io.improbable.keanu.vertices.Vertex;
-
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.stream.Collectors;
+
+import io.improbable.keanu.tensor.Tensor;
+import io.improbable.keanu.vertices.ProxyVertex;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexDictionary;
+import io.improbable.keanu.vertices.VertexLabel;
+
+public class Plate implements VertexDictionary {
 
-public class Plate {
-    private Map<String, Vertex<?>> contents;
+    private Map<VertexLabel, Vertex<?>> contents;
 
     public Plate() {
         this.contents = new HashMap<>();
     }
 
-    public <T extends Vertex<?>> T add(String name, T v) {
-        contents.put(name, v);
+    public <T extends Vertex<?>> T add(T v) throws PlateException {
+        if (v.getLabel() == null) {
+            throw new PlateException("Vertex " + v + " has no label");
+        }
+        VertexLabel label = scoped(v.getLabel());
+        if (contents.containsKey(label)) {
+            throw new IllegalArgumentException("Key " + label + " already exists");
+        }
+        contents.put(label, v);
+        v.setLabel(label);
```

Comment:
That's a good question. However, until we have a way to clone a vertex, I don't see an alternative.

The scoping is necessary because when you put your plates into a BayesNet, you're not allowed duplication of labels.

Reply:
Can we do a little check to se if it has already been added to another plate? And throw with a helpful message if so?

Reply:
done